bindings/go/cpuid.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Supranational LLC
+ * Licensed under the Apache License, Version 2.0, see LICENSE for details.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+int __blst_platform_cap = 0;
+
+#if defined(__x86_64__) || defined(__x86_64) || defined(_M_X64)
+
+# if defined(__GNUC__) || defined(__clang__) || defined(__SUNPRO_C)
+static void __cpuidex(int info[4], int func, int sub)
+{
+    int eax, ebx, ecx, edx;
+
+    asm("cpuid" : "=a"(eax), "=b"(ebx), "=c"(ecx), "=d"(edx)
+                : "a"(func), "c"(sub));
+
+    info[0] = eax;
+    info[1] = ebx;
+    info[2] = ecx;
+    info[3] = edx;
+}
+# else
+#  include <intrin.h>
+# endif
+
+# if defined(__GNUC__) || defined(__clang__)
+__attribute__((constructor))
+# endif
+int __blst_cpuid(void)
+{
+    int info[4], cap = 0;
+
+    __cpuidex(info, 0, 0);
+    if (info[0] > 6) {
+        __cpuidex(info, 7, 0);
+        cap |= (info[1]>>19) & 1; /* ADX */
+        cap |= (info[1]>>28) & 2; /* SHA */
```

Comment:
Noticed a small problem. I think this should be:

```suggestion
        cap |= (info[1]>>29) & 2; /* SHA */
```

Reply:
No, I 'and' it with 2, not 1.

Reply:
I'm confused by your reply, but I think there's a misunderstanding. The SHA feature is the 29th bit of EBX, not the 28th. The 28th bit is `AVX512CD`. See [this page](https://github.com/MicrosoftDocs/cpp-docs/blob/main/docs/intrinsics/cpuid-cpuidex.md?plain=1#L135) for quick reference.

```c
static bool AVX512CD(void) { return CPU_Rep.f_7_EBX_[28]; }
static bool SHA(void) { return CPU_Rep.f_7_EBX_[29]; }
```

Reply:
I aim to set the **second** bit in the `cap`. One way is `(info[1]>>29) & 1) << 1`, and another way is `(info[1]>>28) & 2`.

Reply:
Oh I understand now. Gotcha 👍